### PR TITLE
feat: ClosedLoopEvaluation class framework (PR 2/6)

### PR DESCRIPTION
**Pull request by kosuke55**
## Summary
Part 2 of the stacked closed-loop eval series (stacked on #298). Introduces the ``ClosedLoopEvaluation`` template-method class framework:

- `ClosedLoopEvaluation` (ABC): `discover_jobs` -> `shard_jobs` -> `run_job` -> `build_summary`/`write_artifacts`/`print_summary`, plus a DDP finalize/merge-shards path — `run_distributed()` requires an already-initialized process group and calls `dist.barrier()` before any rank-0 merge, so it can't silently merge a partial result if a rank crashed before writing its shard.
- `FullRouteClosedLoopEvaluation`: concrete segment-wise implementation over every route under an npz_root, using PR1's `render_segment`/`aggregate`. `npz_root` now accepts anything PR1's `resolve_npz_roots` does — a single directory, a `.json` path-list file, or an already-resolved list of roots — not just a single directory: `discover_jobs` merges routes across all of them via PR1's `enumerate_multi_root_routes`, and each route's own source root is threaded through as its job's sidecar dir.
- `RolloutParams`/`ClosedLoopEvalConfig`: shared config dataclasses, including PR1's `drop_objects`/`abort_*` knobs.
- `closed_loop_ddp.py`: just `shard_items` (DDP route sharding) now. Per-rank persistence reuses tier4-main's existing `segments_{rank}.jsonl`/`tdigests_{rank}.jsonl` convention (PR1's sidecar helpers) instead of a separate shard-blob format, so there's nothing left for `finalize_ddp` to write beyond the barrier signal, and `collect_ddp_shards` raises if any rank's file is missing rather than merging a partial result.
- `print_summary`/per-segment logging read PR1's nested `object`/`road_border`/`red_light_violation`/`strong_brake`/`reproducer` metric shape.

Not wired into anything yet — this is the foundation the rest of the stack (site sweep, W&B dashboard, training-loop wiring) builds on.

## Test plan
- [x] `ast.parse` + module import check
- [x] Verified `FullRouteClosedLoopEvaluation` with a real two-site rollout (odaiba + takanawa npz roots passed as a list): both routes ran with correct per-route sidecar dirs and merged into one aggregate summary
- [x] Verified the DDP shard-persistence rewrite: a 2-rank simulation merges correctly, a missing rank's `segments_<rank>.jsonl` fails fast with `RuntimeError`, and calling `run_distributed()` without an initialized process group also fails fast
- [x] Reviewed alongside PR #298 (merged)

### CODE REVIEW

File: scenario_generation/closed_loop_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+        segments_path = self.out_dir / "segments.jsonl"
+        with segments_path.open("w", encoding="utf-8") as fout:
+            for ri, job in enumerate(jobs):
+                assert isinstance(job, FullRouteRouteJob)
+                partial = self.run_job(job, segments_file=fout)
+                merged.rows.extend(partial.rows)
+                merged.video_mp4s.extend(partial.video_mp4s)
+                merged.extras["route_keys"].append(job.route_key)
+                self.on_job_complete(job, partial, ri, len(jobs))
+        return merged
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py:418]: **Bug (please confirm):** under `ddp_world_size > 1`, every rank still opens the shared path `self.out_dir / "segments.jsonl"` here and streams into it.

That races across ranks (last writer wins / truncated / mixed lines). Rank-local work should write `segments_{rank}.jsonl` (matching `tier4-main`), and only the merge step should produce the combined `segments.jsonl`.

This interacts with the `ddp_shards/` path below — even if shard JSON is written later, this early shared write is already unsafe.

File: scenario_generation/closed_loop_ddp.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    *,
+    mode: str,
+    rows: list[dict],
+    video_mp4s: list[Path],
+    elapsed_sec: float,
+    extras: dict[str, Any] | None = None,
+    profile: bool = False,
+) -> Path:
+    """Write a per-rank shard under ``out_dir/ddp_shards/rank_XXX.json``."""
+    shard_dir = out_dir / "ddp_shards"
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_ddp.py]: **Please unify with main’s DDP artifact protocol (please confirm intent).**

`tier4-main` already uses per-rank `segments_{rank}.jsonl` (+ `tdigests_{rank}.jsonl`) and merges those. Introducing `ddp_shards/rank_XXX.json` as a second on-disk protocol means CLI / training / later stacked PRs have to support two merge paths.

Suggestion: keep `shard_items()` (the scheduling helper is fine), but make persistence/merge reuse the existing jsonl convention instead of this blob format.

File: scenario_generation/closed_loop_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+            self.out_dir,
+            self.ddp_rank,
+            mode=self.mode,
+            rows=result.rows,
+            video_mp4s=result.video_mp4s,
+            elapsed_sec=elapsed_sec,
+            extras=self.ddp_shard_extras(result),
+            profile=self.config.profile,
+        )
+        return self.ddp_partial_summary(result, elapsed_sec=elapsed_sec)
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py]: **Related to the protocol split:** `finalize_ddp` writes only via `write_eval_shard` (`ddp_shards/`), while `execute_jobs` may have already clobbered a shared `segments.jsonl`.

If we unify on main’s protocol, this should write/retain `segments_{rank}.jsonl` (and tdigest sidecars when present) and let merge load those, rather than a separate JSON blob tree.

File: scenario_generation/closed_loop_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+            "elapsed_sec": elapsed_sec,
+        }
+
+    def collect_ddp_shards(self, world_size: int) -> JobRunResult:
+        """Load per-rank shard JSON written under ``out_dir/ddp_shards/``."""
+        merged = JobRunResult()
+        for rank in range(world_size):
+            path = self.out_dir / "ddp_shards" / f"rank_{rank:03d}.json"
+            if not path.is_file():
+                continue
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py]: **Bug (please confirm):** missing shard files are silently skipped (`continue`).

If one rank crashes or never reaches `finalize_ddp`, rank 0 can still build a “successful” summary from a subset of routes with no hard failure. Prefer fail-fast (or at least a loud error / explicit failed-rank list) when `rank_{i}.json` / `segments_{i}.jsonl` is absent.

File: scenario_generation/closed_loop_evaluation.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+            import torch.distributed as dist
+        except ImportError:
+            return partial
+        if dist.is_available() and dist.is_initialized():
+            dist.barrier()
+        if self.ddp_rank != 0:
+            return {}
+        if partial.get("ddp_shard"):
+            return self.merge_ddp_shards(self.ddp_world_size)
+        return partial
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py:205]: **Bug / footgun (please confirm):** if `ddp_world_size > 1` but `torch.distributed` is not initialized, the barrier is skipped and rank 0 can still call `merge_ddp_shards()`.

That can merge a partial set of shard files and look like a completed eval. Either require an initialized process group when `world_size > 1`, or refuse to merge unless the caller has already synchronized all ranks.

File: scenario_generation/closed_loop_evaluation.py [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+            f"{summary['n_segments_with_near_miss']}/{n_seg} segments "
+            f"(rate {summary['near_miss_segment_rate']:.4f}), {summary['total_near_miss_steps']} steps"
+        )
+        print(
+            f"global_min_clearance={summary['global_min_clearance']:.3f} m  "
+            f"mean_segment_min_clearance={summary['mean_segment_min_clearance']:.3f} m  "
+            f"mean_segment_mean_clearance={summary['mean_segment_mean_clearance']:.3f} m"
+        )
+        print(f"total_snaps={summary['total_snaps']}  terminated={summary['terminated_counts']}")
+        print(f"videos: per-segment <route>_<start>_<end>.mp4 in {self.out_dir}")
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py:523]: **Rebase note (not the DDP bug, but will break on current `tier4-main`):** this summary printer assumes the pre-nested flat keys (`collision_segment_rate`, `min_clearance`, `n_collision_steps`, …).

After rebasing onto nested metrics + PR1, please switch to `format_summary_lines` / nested blocks, and write rows via `segment_row_for_json` so tdigests stay in the sidecar.

**Review by xtk8532704 — CHANGES_REQUESTED** [review]
## Review summary (PR 2/6)

Thanks for the clean `ClosedLoopEvaluation` template — the overall shape looks good as a foundation for the later site / training PRs.

Before this lands, please **confirm the bugs called out inline**, and please **unify the multi-GPU artifact protocol with current `tier4-main`** rather than introducing a second one.

### Unify DDP artifacts with main

`tier4-main` already shards closed-loop eval via:

- `segments_{rank}.jsonl`
- `tdigests_{rank}.jsonl` (for clearance p5 merge)
- parent / rank-0 merge through the existing helpers in `closed_loop_eval.py` / `valid_predictor_closed_loop.py`

This PR adds a parallel protocol under `ddp_shards/rank_XXX.json`. That will force PR5/PR6 and the existing CLI onto a forked path. Prefer keeping the class framework, but making `finalize_ddp` / merge speak the existing jsonl protocol (and dropping `ddp_shards/` as the standard).

### Blocking issues called out inline

1. **Multi-rank race on `segments.jsonl`** — under `world_size > 1`, every rank still opens the same `segments.jsonl` in `execute_jobs` before writing `ddp_shards/`.
2. **Second DDP artifact protocol** — `write_eval_shard` / `ddp_shards/` duplicates main’s jsonl sharding.
3. **Silent missing shards** — `collect_ddp_shards` skips absent rank files and can under-count without failing.
4. **`run_distributed` without an initialized process group** — barrier may be skipped while rank 0 still merges, so incomplete shards can be treated as final.

### Also after rebasing onto current main (+ PR1)

Once this sits on nested-metrics `tier4-main`, `print_summary` / row writing / `RolloutParams` will also need to align with `segment_row_for_json`, `format_summary_lines`, `yaw_gate`, and `strong_brake_mps2`. Happy to re-review after the DDP unify + bugfixes.

Please reply on the inline threads confirming whether you agree with each finding.

**powered by cursor**

**Comment by kosuke55** [comment]
rebased on tier4-main

File: scenario_generation/closed_loop_evaluation.py (hunk 1 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py:418]: fixed in ab8e3ce213a25487dc2a7537a3bf13550daab65f

File: scenario_generation/closed_loop_ddp.py (hunk 2 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_ddp.py]: fixed in ab8e3ce213a25487dc2a7537a3bf13550daab65f

File: scenario_generation/closed_loop_evaluation.py (hunk 3 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py]: fixed in ab8e3ce213a25487dc2a7537a3bf13550daab65f

File: scenario_generation/closed_loop_evaluation.py (hunk 4 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py]: fixed in ab8e3ce213a25487dc2a7537a3bf13550daab65f

File: scenario_generation/closed_loop_evaluation.py (hunk 5 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py:205]: fixed in eaa213dbd7a400466ef5d1fa862a5afc468619b2

**Comment by kosuke55** [comment]
@xtk8532704 
Thanks for the review! I’ve made the changes, so please check it out.

**Review by xtk8532704 — COMMENTED** [review]
Thanks for addressing the previous review comments — the DDP path looks good now (per-rank `segments_{rank}.jsonl` / `tdigests_{rank}.jsonl`, fail-fast on missing shards, and requiring an initialized process group before merge). Nested metrics printing / `segment_row_for_json` also look aligned.

One leftover question from the rebase note (not a hard blocker): do `yaw_gate` and `strong_brake_mps2` need to be plumbed into `RolloutParams` / `render_kwargs()` (and passed through to `aggregate(...)`) for parity with main's `run_closed_loop_eval`, or are you intentionally leaving them at `render_segment` / `aggregate` defaults for now?

File: scenario_generation/closed_loop_evaluation.py [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+    # above the unstick_* knobs above: unstick snaps the ego back onto GT to let a merely-stuck
+    # rollout continue; abort gives up on a rollout unstick can't save. 0 = disabled for either
+    # knob. See ``reproducer_rollout.render_segment`` for the exact trigger condition.
+    abort_deviation_m: float = 50.0
+    abort_after: int = 30
+    abort_max_snaps: int = 0
+    # Empty-world ablation: zero neighbor_agents_past + static_objects each step (map kept) —
+    # separates "reacts badly to traffic" from "can't follow the route/map". collision/near-miss
+    # are 0 by construction when this is set.
+    drop_objects: bool = False
```
Comment (xtk8532704) [inline scenario_generation/closed_loop_evaluation.py:67]: Question: should `yaw_gate` and `strong_brake_mps2` also live on `RolloutParams` (and be forwarded via `render_kwargs()` / into `aggregate(...)`)? Defaults currently match main (`True` / `-2.5`), so behavior is fine as-is — just checking whether you want this knob surface here for later CLI / training wiring.

Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py:67]: oh sorry I forgot adding them. added in a197703e75efc67e98c80d8a2c2f7969ab6139ab

**Review by xtk8532704 — APPROVED** [review]
LGTM!

File: scenario_generation/closed_loop_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+            )
+        rows = load_segment_rows_with_tdigests(self.out_dir, shard_glob="segments_*.jsonl")
+        video_mp4s = sorted(self.out_dir.glob("*.mp4"))
+        return JobRunResult(rows=rows, video_mp4s=video_mp4s)
+
+    def merge_ddp_shards(self, world_size: int) -> dict:
+        """Rank-0: merge shard files, persist artifacts, and return the final summary."""
+        result = self.collect_ddp_shards(world_size)
+        self.prepare_ddp_merge_artifacts(result)
+        summary = self.build_summary(result, elapsed_sec=result.elapsed_sec)
```
Comment (go-sakayori) [inline scenario_generation/closed_loop_evaluation.py]: `result.elapsed_sec` here is always 0.0. `collect_ddp_shards` never restores it, so the merged summary prints "in 0.0s" and persists elapsed_sec: 0.0. Consider timing the run in run_distributed

**Review by go-sakayori — CHANGES_REQUESTED** [review]
Please also rebase to tier4-main.

File: scenario_generation/closed_loop_evaluation.py (hunk 8 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py]: oh thank you fixed
https://github.com/tier4/Diffusion-Planner/pull/299/commits/2e6853a08ae1c26fdf6588f6bc4138ffc51213a6

File: scenario_generation/closed_loop_evaluation.py [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+
+
+@dataclass
+class JobRunResult:
+    """Accumulated rows + artifacts from one or more jobs on this rank."""
+
+    rows: list[dict] = field(default_factory=list)
+    video_mp4s: list[Path] = field(default_factory=list)
+    extras: dict[str, Any] = field(default_factory=dict)
+    elapsed_sec: float = 0.0
```
Comment (go-sakayori) [inline scenario_generation/closed_loop_evaluation.py]: nit: After `be92660f`, nothing reads `JobRunResult.elapsed_sec` anymore. Harmless, so not a blocker, but consider dropping this field in a follow-up to avoid suggesting it carries meaningful timing data (it's always 0.0).

**Review by go-sakayori — APPROVED** [review]

File: scenario_generation/closed_loop_evaluation.py (hunk 9 above)
Comment (kosuke55) [reply scenario_generation/closed_loop_evaluation.py]: thanks! drop it
https://github.com/tier4/Diffusion-Planner/pull/299/changes/9e22db72510d65f0511a99ed760e4d0d4842ed00
